### DISCUSSION TITLE
Store values in the scope of a web request

### DISCUSSION THREAD

**Discussion by laped** [vikejs/vike#941] [Discussion] [answered]
I'm looking for a way to store values in the scope of a web request, for example a http header value from the incoming request or a traceId/transactionId for logging/distributed tracing purposes and access those values within pages or imported modules that are all part of the request.

I don't want to propagate values manually through all the function calls as arguments and the only way I've found so far that should be able to handle this, is AsyncLocalStorage.

But when I use it with vite-plugin-ssr (or vite, I don't know specifically which one causes this), the context is lost when rendering a page serverside. See this example: https://github.com/laped/vite-ssr-asynclocalstorage

When running this code, this is what I see in the console log:

```
nodejs renderPage:before - store:  Map(1) { 'traceId' => 'Test123' }
_default.page.server - store:  undefined
index.page - store:  undefined
nodejs renderPage:after - store:  Map(1) { 'traceId' => 'Test123' }
```

So from the express middleware to the endpoint, that context is available and also before/after renderPage - but not within the server renderer or the page itself, the store is just undefined.

Any idea why that is or how else I could go about doing something like this?

**Comment by brillout** [comment, marked as answer]
Neither VPS nor Vite interferes with `AsyncLocalStorage`, so I'm inclined to think that the problem is that you aren't using it correctly.

You may need to use `globalThis` to share information between modules.

**Reply by laped** [reply]
From my understanding of the docs and other examples I've seen of using AsyncLocalStorage, I don't see my implementation being any different. I create a new instance of the AsyncLocalStorage in a module and then export it, which to my understanding results in a singleton when imported into other modules - as they would all get the same instance.

I've updated my example above and added a random generated moduleId to the AsyncLocalStorage module, to show which module instance is being loaded.

I've also added a different express endpoint, which calls and awaits an async function from another module (testModule). That module imports my AsyncLocalStorage module also, reads and console logs those values from the storage. This is what the console log looks like when the /testModule endpoint is called:

```
nodejs testModule:before - store:  Map(1) { 'traceId' => 'Test123' } 58299716504464993427
testModule - store:  Map(1) { 'traceId' => 'Test123' } 58299716504464993427
nodejs testModule:after - store:  Map(1) { 'traceId' => 'Test123' } 58299716504464993427
```

_The big number at the end of the log line is the moduleId._

Then immediately afterwards I call the VPS/Vite endpoint (just the root http://localhost:3000/) and this is the log output:

```
nodejs renderPage:before - store:  Map(1) { 'traceId' => 'Test123' } 58299716504464993427
_default.page.server - store:  undefined 54472144216184914534
index.page - store:  undefined 54472144216184914534
nodejs renderPage:after - store:  Map(1) { 'traceId' => 'Test123' } 58299716504464993427
```

As you can see, even across multiple requests, the moduleId stays the same, but not inside VPS/Vite page/context where the moduleId is now different and the store is undefined. So I think something incorrect is happening inside either VPS renderPage or in vite dev middleware.

**Reply by brillout** [reply]
Tbh I didn't look it your reproduction. Because I strongly suspect it has nothing to do with VPS nor Vite. AFAIK async hooks are completely orthogonal. But feel free to keep me updated if you find the problem.

**Reply by laped** [reply]
Storing the instance of AsyncLocalStorage in globalThis worked, so thanks for that suggestion!

But I might look into this a bit more, because I think it should just work out of the box without globalThis. I will let you know if I find any problems.

**Reply by brillout** [reply]
👍 FYI https://github.com/sponsors/brillout - it's prob a bit premature but in case you can (eventually) convince your company.
